### PR TITLE
MM-54416: Channel Bookmarks Web UI

### PR DESCRIPTION
**Pull request by calebroseland**
#### Summary
Feature complete with fully functional files and links.

While there are still a few outstanding items, review and smoke testing can be done if the test server successfully starts up.

TODO: 
- [x] sync with upstream
- [x] 50x cap/limit
- [x] show error messages during file upload/save flows
- [x] crash/issue with plugin hooks while previewing files
- [x] de-lint
- [x] Restrict opengraph api according to current Link Preview security/semantics
- [x] E2E/unit tests

Edge cases/issues discovered:
- ~cancel pending open graph network request(s) on link changed~ done
- ~existing opengraph mechanism doesn’t pick up all website icons, nor plain site titles—add fallback parsing to pick up non opengraph titles and icons~ https://mattermost.atlassian.net/browse/MM-56761
- ~display name is unbound (e.g. 30KB worth)~ https://mattermost.atlassian.net/browse/MM-56775
- ~after a file bookmark is deleted, the file still accessible via Channel Files (search)~ https://mattermost.atlassian.net/browse/MM-56774
- handle delete/updates optimistically to prevent stale UI state on slow connections
- bookmark deleted by someone else while editing causes error when they go to save it as deleted bookmarks can't be undeleted to my knowledge

Bugs: 

Deferred:
- reordering
- fetch optimization using `since` logic, instead of always fetching all
- ~~integrated header [MM-56324](https://mattermost.atlassian.net/browse/MM-56324), to avoid UX frustration when channel header hover covers over bookmarks and keeps reappearing on mouse-out-in-out-in~~

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-54416: 
- https://mattermost.atlassian.net/browse/MM-54430
- https://mattermost.atlassian.net/browse/MM-54431
- https://mattermost.atlassian.net/browse/MM-56285
- https://mattermost.atlassian.net/browse/MM-56287
- https://mattermost.atlassian.net/browse/MM-56325
- https://mattermost.atlassian.net/browse/MM-56350
- https://mattermost.atlassian.net/browse/MM-54427
- https://mattermost.atlassian.net/browse/MM-56647
#### Screenshots
Current status: 

https://github.com/mattermost/mattermost/assets/11724372/601b88ce-f8f0-4005-87bb-11814de5bae1

https://github.com/mattermost/mattermost/assets/11724372/6f8ab021-519d-4ea3-9bd3-2f202f03213a

https://github.com/mattermost/mattermost/assets/11724372/748db6a8-45ed-4b21-b7ee-4479ad189672

#### Release Note

```release-note
Channel Bookmarks UI
```

[MM-56324]: https://mattermost.atlassian.net/browse/MM-56324?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webapp/platform/client/src/client4.ts
Comment (enahum) [inline webapp/platform/client/src/client4.ts:1567]: any reason why not to include `bookmarks_since` ?

File: webapp/platform/client/src/client4.ts
Comment (enahum) [inline webapp/platform/client/src/client4.ts:1719]: same question as above

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/channel_bookmarks.ts
Comment (enahum) [inline webapp/channels/src/packages/mattermost-redux/src/selectors/entities/channel_bookmarks.ts:20]: I fail to see why the `?` is needed as if no bookmarks the selector seems to be returning an empty object `EMPTY_BOOKMARKS`

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/channel_bookmarks.ts
Comment (enahum) [inline webapp/channels/src/packages/mattermost-redux/src/reducers/entities/channel_bookmarks.ts:59]: at some point, when using the `bookmarks_since`, you should check if some bookmarks should be removed from state.

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/channel_bookmarks.ts
Comment (enahum) [inline webapp/channels/src/packages/mattermost-redux/src/reducers/entities/channel_bookmarks.ts:73]: same as above

File: webapp/channels/src/packages/mattermost-redux/src/actions/channel_bookmarks.ts
Comment (enahum) [inline webapp/channels/src/packages/mattermost-redux/src/actions/channel_bookmarks.ts:57]: when you edit a bookmark created by another user, you will receive a `deleted` property with a bookmark that got deleted. in this scenario you have a bookmark deleted and created. You should delete the bookmark in state and add / update the other one (the one you are currently processing)

File: webapp/channels/src/components/channel_bookmarks/bookmark_delete_modal.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/bookmark_delete_modal.tsx:46]: is this intended?

File: webapp/channels/src/components/channel_bookmarks/bookmark_dot_menu.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/bookmark_dot_menu.tsx:42]: don't forget about this

File: webapp/platform/client/src/client4.ts
Comment (enahum) [inline webapp/platform/client/src/client4.ts:1948]: all of these should include the WebSocket connection ID to exclude the same client from receiving the WS events

File: webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx:1]: OpenGraph and file missing.. I guess it has not been completed yet

**Review by enahum — CHANGES_REQUESTED** [review]

**Comment by esethna** [comment]
FYI @crspeller Scott and I added you to review here. @calebroseland please ping reviewers when this is ready for re-review

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/channel_bookmarks.ts
Comment (calebroseland) [reply webapp/channels/src/packages/mattermost-redux/src/selectors/entities/channel_bookmarks.ts:20]: Good point. Was mainly from habit, but not indeed needed here.

File: webapp/channels/src/components/channel_bookmarks/bookmark_delete_modal.tsx
Comment (calebroseland) [reply webapp/channels/src/components/channel_bookmarks/bookmark_delete_modal.tsx]: The empty func is needed for the cancel button to show

File: webapp/channels/src/packages/mattermost-redux/src/actions/channel_bookmarks.ts
Comment (calebroseland) [reply webapp/channels/src/packages/mattermost-redux/src/actions/channel_bookmarks.ts]: Yep, that was in progress at the time, done now

File: server/channels/app/opengraph.go
Comment (crspeller) [inline server/channels/app/opengraph.go:57]: This is kinda confusing here. Maybe we should just read all here and create a `bytes.Reader` for `og.ProcessHTML`? We read all in `parseLinkFavicon` anyway.

**Review by crspeller — CHANGES_REQUESTED** [review]

**Comment by esethna** [comment]
@calebroseland can you help get the test server up and running so we can start UX and QA review?

**Comment by calebroseland** [comment]
@esethna I haven’t been able to get it spun up on this PR for some reason, if it continues to have problems I’ll try to get a dedicated server running it as a backup after lunch here.

**Comment by esethna** [comment]
Thanks @calebroseland, can you also help provide an update on the remaining to dos?

**Comment by calebroseland** [comment]
> Thanks @calebroseland, can you also help provide an update on the remaining to dos?

I will do that this afternoon, as well 👍

**Comment by calebroseland** [comment]
Alright, https://mattermost-pr-26075.test.mattermost.cloud/pr26075/ is up under #26075. File errors are now handled, and the 50 per-channel is preemptively enforced. Currently working on unit/e2e tests, tidying things up, and review comments.

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Great work on this @calebroseland. This is solid. I went through the UI/UX and have some minor details I'd like to call out and one more significant challenge with the horizontal scrolling.

1. For the `+ Add bookmark` button:
    - Can we update the + icon to be 16x16
    - Can we update the padding on this button to `padding: 4px 12px 4px 6px;`
    - For comparison, here's implementation vs. design
    image
    - Note: this would change the small icon-only version of this button that appears anchored to the right, so be sure to be specific to the version of the button that has the label.

2. There looks to be a container around the `+ Add bookmark` that is causing it to look misaligned with the content above in the channel header. `PlusButtonContainer` has `padding: 0 1rem;`. I think this padding can be removed.
    image

3. For the fallback icon (when a link can't pull a favicon), Asaad had the `icon-book-outline` used instead on his mobile designs. Can we do the same for the webapp?
image

4. For link bookmarks, can we update the icon used in the `Open` menu item? Can we use `icon-open-in-new` instead of `icon-arrow-expand`?
    image

4. For link bookmarks, is it possible to omit the requirement for `https://` or `http://`? From a UX perspective it's an annoying set of characters to have to type. It would be nice if users could just enter `mattermost.com` instead. Is that doable?

6. Links don't seem to be pulling in the favicon right now for me. I tried adding `https://www.mattermost.com` and it didn't use the favicon.

7. For the overflow when there are too many bookmarks to fit, I think we may need to rethink the approach here because the horizontal scrollbar is too visually jarring. When items overflow, I would suggest the + button changes to a ••• button which opens a menu showing the overflow items and the add bookmark option. Like this:
    image
    - this is what I propose
    image
    - I understand this may not be easy to accomplish since it would require dynamic calculations to show/hide bookmarks in the overflow menu depending on screenwidth, but the horizontal scrollbar isn't going to work as-is IMO. 
    - Let's chat about this if what I'm proposing isn't feasible. We'll need to figure out a solution either way

8. For the error message that displays under the link URL input field, the icon is looking slightly misaligned - it's sitting too high by the looks of it.
    image

9. For the help text that shows initially under the link URL input field, can we adjust the color to `color: rgba(var(--center-channel-color-rgb), 0.72);`?

10. I also noticed the overall height of the bookmarks bar is measuring 33px for me, but the designed height was 36px. I'm not sure what's causing that thought because the chips are the right height (24px) and the padding is right (6px). 24+6+6 = 36. Not sure why I'm getting 33px. Obviously, this is really nit-picky and not a big deal.

**Comment by lindy65** [comment]
Thanks @calebroseland - I have done a first QA pass. Here are a few questions / observations:

1. Should it be possible to save .exe and .msi files as bookmarks? I'm thinking this might be a security risk?

2. Not sure whether this is related to your PR... I checked on `master.test.mattermost.com` test server and marking posts as unread / clicking on "Jump to Unreads" is working as expected. I'm testing how the [bookmarks bar displays with other banners](https://mattermost.atlassian.net/wiki/spaces/WD/pages/2471591944/UX+Documentation+-+Bookmarks+Bar#How-it-displays-with-other-banners) and tried to mark some posts as unread in Town Square channel so I could go back into the channel to see the "Jump to Unreads" banner.  
    -  Marking posts as unread does not seem to be working consistently on the spinwick
    -  When marking posts as unread _does_ work and I go back into the channel and click on "Jump to Unreads", the channel does not scroll up to the first unread post
    

https://github.com/mattermost/mattermost/assets/13020833/e2a3dd22-38b1-4319-87e4-60d83fa64421



https://github.com/mattermost/mattermost/assets/13020833/9a1281fb-767e-4b63-9dc7-1e61b16649a6


3. Should I be able to press Enter on my keyboard to add the bookmark (rather than having to click on "Add bookmark" each time?
image

4. I tested the limit of 50 bookmarks per channel which is working as expected.  Just saying... It would be nice to have a counter of the number of bookmarks you have in a channel although I realise there is a tooltip that appears "Cannot add bookmark. Limit reached" once you have 50 bookmarks in the channel

5. Not sure what @matthewbirtch thinks but I feel the text on the bookmark limit reached tooltip would look better if "Limit reached" was all on the 2nd line?

image

7. Not able to search for an emoji in click on elipsis to the right of bookmark name > Edit > click on dropdown to the left of bookmark title > click on "Search emojis" in emoji picker.  Observed: "Search emojis" field does not activate/focus. Expected: "Search emojis" field is focused when emoji picker opens.

image

8. I'm not able to [re-order bookmarks](https://mattermost.atlassian.net/wiki/spaces/WD/pages/2471591944/UX+Documentation+-+Bookmarks+Bar#Reorder-Bookmarks) (drag 'n drop).  Is this not yet implemented?

9. Search files in bookmarks:  This is working as expected.  I have a question relating to the 3-dot menu that is visible on the file in the RHS when hovering over the file - see image below:

image

10. Similarly, when clicking on "Copy link" in the above image, then pasting the link into a channel, the URL shows `undefined` at the end. Should the "Copy link" option also be removed?
    - `https://mattermost-pr-26075.test.mattermost.cloud/pr26075/pl/undefined`

11. When filtering bookmark file search results by type (e.g. documents), a "loading" icon is displayed below the search result.  In the image below, there should only be one result but the loading icon continues to display

image

12. Being able to "copy link" on bookmarks that are files doesn't make sense to me as the file needs to be downloaded to view it.  Copying and pasting the link into a channel, then clicking on it results in a blank loading page.  Should the "copy link" option be removed from bookmarks that are files?

image

image

**Comment by matthewbirtch** [comment]
> 5\. Not sure what @matthewbirtch thinks but I feel the text on the bookmark limit reached tooltip would look better if "Limit reached" was all on the 2nd line?

@lindy65 I'd like to update the copy to include what the maximum limit is.

`Cannot add more bookmarks. The maximum number of bookmarks is 50`. 

Also, we should display the + button as disabled in this state and change the mouse cursor on the button to `cursor: not-allowed`.

**Comment by calebroseland** [comment]
cc @esethna

I've addressed a number of the UX/QA review points and am currently working on negating duplicate opengraph requests, tests, and other more minor review items.

Stepping away for the time being, but I'll be on later tonight to push some pending changes, and update in more detail on which items I can address now and which ones we'll have to open tickets to do later

In short: this should be ready for a final round of reviews late this weekend or early Monday.

**Comment by esethna** [comment]
Thanks @calebroseland, please ping reviewers when needed to re-engage.

**Comment by calebroseland** [comment]
- `Matt 1–10`: All done, except 6 and 7 which have followup tickets. Can discuss Monday.
- `Lindy 1`: .exe, .msi, and such are supported in post attachments. Both Post and Bookmark attachments use the same upload pipeline.
- `Lindy 2`: Appears to be working correctly. (Videos appear to be trying to mark the own user posts as unread.)
- `Lindy 3`: Done, `enter` now saves
- `Lindy 4`: Let's discuss Monday; needs UX input. Will be followup if anything to be done.
- `Lindy 5`: Text changed, disabled button styling corrected
- `Lindy 8`: Followup ticket
- `Lindy 6,7,9,10,11`—in-progress
- `Lindy 12`: Let's discuss Monday; propose swapping `Copy Link` with `Get Public Link` for files

**Comment by matthewbirtch** [comment]
> * `Matt 1–10`: All done, except 6 and 7 which have followup tickets. Can discuss Monday.

Great work Caleb. Verified that you covered the issues I identified (other than the ones you've got follow-ups for).

**Review by matthewbirtch — APPROVED** [review]
Let me know when you've got PRs for those follow-up tickets as well/

**Comment by calebroseland** [comment]
> Let me know when you've got PRs for those follow-up tickets as well.

We may be able to fit 6 in, but 7 probably won't be a possibility—it's a good solution, but there's not enough time to do it now.

**Comment by matthewbirtch** [comment]
> > Let me know when you've got PRs for those follow-up tickets as well.
> 
> 
> 
> We may be able to fit 6 in, but 7 probably won't be a possibility—it's a good solution, but there's not enough time to do it now.

@calebroseland can we create a follow up ticket still? And maybe we can make it a help-wanted ticket if we can't fit it in.

**Comment by lindy65** [comment]
> * `Lindy 1`: .exe, .msi, and such are supported in post attachments. Both Post and Bookmark attachments use the same upload pipeline.

Thanks 👍 

> * `Lindy 2`: Appears to be working correctly. (Videos appear to be trying to mark the own user posts as unread.)

Oh yeah, of course! 🤦‍♀️ 

> * `Lindy 3`: Done, `enter` now saves

Thanks, tested, working as expected

> * `Lindy 4`: Let's discuss Monday; needs UX input. Will be followup if anything to be done.

👍 

> * `Lindy 5`: Text changed, disabled button styling corrected

Thanks, tested and this is fixed/working as expected

> * `Lindy 8`: Followup ticket

👍 

> * `Lindy 6,7,9,10,11`—in-progress

👍 

> * `Lindy 12`: Let's discuss Monday; propose swapping `Copy Link` with `Get Public Link` for files

Sounds good 👍

File: server/channels/app/opengraph.go
Comment (calebroseland) [reply server/channels/app/opengraph.go]: Done

File: server/channels/api4/channel_bookmark.go
Comment (calebroseland) [reply server/channels/api4/channel_bookmark.go:409]: Put `GetOpenGraphMetadata` behind  `ServiceSettings.EnableLinkPreviews` and `isLinkAllowedForPreview`, which is what `parseLinkMetadata` is secured by, I believe. 

cc @jupenur would you mind taking a look at this?

**Comment by calebroseland** [comment]
@jupenur cc @esarafianou the main item in question on this one is `fetchOpenGraph`—https://github.com/mattermost/mattermost/pull/25889/files#r1478357215

**Review by matthewbirtch — APPROVED** [review]
Looking good to me @calebroseland. Thanks for creating the follow-up tickets.

File: webapp/channels/src/components/channel_bookmarks/utils.ts
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/utils.ts:1]: perhaps this file should be called `hooks` instead

File: webapp/channels/src/components/channel_bookmarks/create_modal_name_input.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/create_modal_name_input.tsx:26]:
```suggestion
    fileInfo?: FileInfo;
    imageUrl?: string;
    emoji?: string;
```

File: webapp/channels/src/components/channel_bookmarks/create_modal_name_input.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/create_modal_name_input.tsx:29]:
```suggestion
    placeholder?: string;
    displayName?: string;
```

File: webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx:163]: 1/5 not use of Client4 directly

File: webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx:420]: this functions get recreated every time that will cause `GenericModal` to re-render, specially when uploading a file as the progress updates.

File: webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx:185]: in theory `channelId` is missing

File: webapp/channels/src/components/channel_bookmarks/bookmark_item.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/bookmark_item.tsx:95]: this could show a chip with a dot menu without any item as the `if/else` if statement could not be met at some point

File: webapp/channels/src/components/channel_bookmarks/bookmark_icon.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/bookmark_icon.tsx:44]: what if this fails? should we show the bookmark outline icon when that happens ?

File: webapp/channels/src/components/channel_bookmarks/bookmark_dot_menu.tsx
Comment (enahum) [inline webapp/channels/src/components/channel_bookmarks/bookmark_dot_menu.tsx:107]:
```suggestion
    }, [bookmark.file_id, dispatch]);
```

File: webapp/channels/src/actions/channel_bookmarks.ts
Comment (enahum) [inline webapp/channels/src/actions/channel_bookmarks.ts:37]: is there any reason why we are not using the `bookmarks_since` here ?

**Review by enahum — APPROVED** [review]
non blocking comments, overall looking good, I'll leave it to you if you want to make the changes

**Comment by enahum** [comment]
@calebroseland you need to resolve conflicts on this PR

**Comment by lindy65** [comment]
Hi @calebroseland - I'm wanting to do final review on the points you've resolved but when going to the [test server](https://mattermost-pr-26075.test.mattermost.cloud/pr26075/), it seems to be a fresh server.  I added an enterprise licence but still don't see the Bookmarks bar below the channel header.  I've check Permissions which are correct.  Is the feature behind a flag now? (I've checked feature flags in the Sys Cons but don't see Bookmarks there?)
Perhaps all this is due to the conflicts Elias mentioned above?

File: server/channels/api4/channel_bookmark.go
Comment (calebroseland) [reply server/channels/api4/channel_bookmark.go:409]: It would've been good to keep it as the same implementation, and to make requests client-side to reduce a small bit of network activity on the server, but `fetchRaw` throws CORS from my testing

```
Access to fetch at 'https://visualstudio.com/' from origin 'https://cr-dev.test.mattermost.cloud' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request.
```

File: webapp/platform/types/src/client4.ts
Comment (calebroseland) [reply webapp/platform/types/src/client4.ts:23]: AbortController signal

File: webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx
Comment (calebroseland) [reply webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx]: added

File: webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx
Comment (calebroseland) [reply webapp/channels/src/components/channel_bookmarks/channel_bookmarks_create_modal.tsx:419]: added useCallback

**Comment by lindy65** [comment]
> Hi @lindy65, I'm not sure what happened with the test server. I had to re-add the label yesterday because it wouldn't update after some pushes. When I re-added it, it did fire up, so I thought it was fine—but I checked the version and https://mattermost-pr-26075.test.mattermost.cloud/pr26075/ was running an old or stock build somehow for some odd reason (it didn't appear to have any of the channel bookmarks code). I was getting crashing errors just searching for the feature flags page in system console—just searching `fea` caused system console to crash.
> 
> CleanShot 2024-02-06 at 09 35 34@2x

Hi @calebroseland - I'm seeing the same on the test server spun up.  Today when I went to the server, it was a Team Edition server so I upgraded to Enterprise and restarted the server.  The Enterprise licence was added.

I still do not see any bookmarks bar feature on this server though.  I managed to access the feature flags section of the sys console by typing `flag` and do not see bookmarks bar in the feature flags.  Btw, I also get a crash if I type `fea` into the sys console search bar.  I'm redirected to the page below and see errors in the js.console

image

I updated the branch on https://github.com/mattermost/mattermost/pull/26075 and the spinwick updated and now has the bookmarks bar code on it so I will go ahead and test the fixes you made for the issues I reported in my previous review.

image

image

**Comment by lindy65** [comment]
Hi @calebroseland 

Re-reviewing the fixes you implemented...

I can still reproduce the loading animation in the RHS when only 1 result is found (and should only be one result found). I can't repro this if there is more than one result found and filtered on "Files" in the RHS.  Seeing the loading animation when it does repro seems to be only on the first time search of the term

image

Other than this issue, the fixes / updates you made are working as expected.

File: server/i18n/en.json
Comment (calebroseland) [reply server/i18n/en.json]: added

**Comment by lindy65** [comment]
> @lindy65 thanks for the find, I think the remaining search issue seems like a corner case and may not be related to these code changes specifically. I'd propose we test on community after this is merged and create a follow-up ticket if still needed.

Sounds good to me, thanks @esethna - I'll go ahead and QA-approve this PR then

**Review by lindy65 — APPROVED** [review]
Thanks @calebroseland - great job :) I'll test the loading animation issue post merge and create a follow-up ticket if I can still repro then.

**Review by crspeller — APPROVED** [review]

File: server/channels/api4/channel_bookmark.go
Comment (jupenur) [inline server/channels/api4/channel_bookmark.go:21]: We had this before but removed it because of security issues. There needs to be a thorough discussion of what needs to be done in order to reintroduce it. And if/when we do reintroduce it, I'd prefer we use the original implementation (with required improvements) instead of implementing a new API just for this use case specifically.

https://github.com/mattermost/mattermost/pull/24177

**Review by jupenur — CHANGES_REQUESTED** [review]
Just getting started with the review, more to come.

**Review by jupenur — CHANGES_REQUESTED** [review]

**Review by jupenur — COMMENTED** [review]
There's a bug here that causes files attached channel bookmarks not to be deleted when you delete the bookmark: https://github.com/mattermost/mattermost/blob/6773d13deee47a437db07e74b0675992cfa912dd/server/channels/store/sqlstore/channel_bookmark_store.go#L320

The method first updates the `DeleteAt` value in the `ChannelBookmarks` table, then attempts to update `FileInfo` rows that match a non-deleted `ChannelBookmarks` row.

**Comment by agnivade** [comment]
I assume this PR included DB migrations at some point but now it's not there? Maybe that's why a review to Server platform is there.

**Comment by jupenur** [comment]
@agnivade I believe Server platform review was for the new OpenGraph API. That got dropped because of security concerns.

**Review by jupenur — APPROVED** [review]
The webapp UI stuff all looks good. Just the two comments above that I'd like addressed before merging.

File: server/public/model/feature_flags.go
Comment (calebroseland) [reply server/public/model/feature_flags.go]: Yes, back to disabled by default now

**Comment by calebroseland** [comment]
> There's a bug here that causes files attached channel bookmarks not to be deleted when you delete the bookmark:
> 
> https://github.com/mattermost/mattermost/blob/6773d13deee47a437db07e74b0675992cfa912dd/server/channels/store/sqlstore/channel_bookmark_store.go#L320
> 
> The method first updates the `DeleteAt` value in the `ChannelBookmarks` table, then attempts to update `FileInfo` rows that match a non-deleted `ChannelBookmarks` row.

@enahum is this deleteat safe to remove from `fileIdQuery`? 

@jupenur since this PR no longer touches the backend and the feature flag is still default-off, I would prefer to handle this via a separate ticket.

**Pull request merged**
